dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedNodeSelectionStrategyChannel.java

```diff
@@ -92,25 +116,101 @@
 
         // TODO(dfox): P2C optimization when we have high number of nodes to save CPU?
         // http://www.eecs.harvard.edu/~michaelm/NEWWORK/postscripts/twosurvey.pdf
-        SortableChannel[] sortedChannels = sortByScore(preShuffled);
+        ChannelStats[] sortableChannels = computeScores(preShuffled);
+        Arrays.sort(sortableChannels, BY_SCORE);
 
-        for (SortableChannel channel : sortedChannels) {
+        for (ChannelStats channel : sortableChannels) {
             Optional<ListenableFuture<Response>> maybe = channel.delegate.maybeExecute(endpoint, request);
             if (maybe.isPresent()) {
+                maybeSampleAllChannelRttsNonBlocking();
                 return maybe;
             }
         }
 
         return Optional.empty();
     }
 
-    private static SortableChannel[] sortByScore(List<MutableChannelWithStats> preShuffled) {
-        SortableChannel[] sorted = new SortableChannel[preShuffled.size()];
+    private static ChannelStats[] computeScores(List<MutableChannelWithStats> preShuffled) {
+        ChannelStats[] snapshotArray = new ChannelStats[preShuffled.size()];
+
+        long bestRttNanos = Long.MAX_VALUE;
+        long worstRttNanos = 0;
         for (int i = 0; i < preShuffled.size(); i++) {
-            sorted[i] = preShuffled.get(i).computeScore();
+            ChannelStats snapshot = preShuffled.get(i).getSnapshot();
+            snapshotArray[i] = snapshot;
+
+            if (snapshot.rttNanos != Long.MAX_VALUE) {
+                bestRttNanos = Math.min(bestRttNanos, snapshot.rttNanos);
+                worstRttNanos = Math.max(worstRttNanos, snapshot.rttNanos);
+            }
+        }
+
+        // Latency (rtt) is measured in nanos, which is a tricky unit to include in our 'score' because adding it
+        // would dominate all the other data (which has the unit of 'num requests'). To avoid the need for a conversion
+        // fudge-factor, we instead figure out where each rtt lies on the spectrum from bestRttNanos to worstRttNanos,
+        // with 0 being best and 1 being worst. This ensures that if several nodes are all within the same AZ and
+        // can return in ~1 ms but others return in ~5ms, the 1ms nodes will all have a similar rttScore (near zero).
+        // Note, this can only be computed when we have all the snapshots in front of us.
+        long rttRange = worstRttNanos - bestRttNanos;
+        if (bestRttNanos != Long.MAX_VALUE && worstRttNanos != 0 && rttRange > 0) {
```

Comment:
perhaps a negative sentinel value, extracted to a static field to make the code easier to read? `bestRttNanos != UNKNOWN_RTT`

Reply:
So this is actually the result of doing the `min` and `max` things above.  I tried using streams to compute these:
```java
Arrays.stream(snapshotArray)
                .filter(snapshot -> snapshot.rttNanos.isPresent())
                .mapToLong(snapshot -> snapshot.rttNanos.getAsLong())
                .summaryStatistics();
```

but the summaryStatistics _also_ gives you a Long.MIN_VALUE/MAX_VALUE if there is no inputs.